design/authenticators/authn_gcp/authn_gcp_solution_design.md

```diff
@@ -0,0 +1,312 @@
+- [Logs](#logs)
+  * [Audit](#audit)
+- [Documentation](#documentation)
+- [Open questions](#open-questions)
+- [Implementation plan](#implementation-plan)
+
+## Glossary
+
+| **Term** | **Description** |
+|----------|-----------------|
+|  GCP| Google Cloud Platform                 |
+|  Google compute engine (GCE)| Virtual machine in google (similar to AWS EC2)                  |
+|  Environment-provided service account| An identity for accessing private data on behalf of a service account inside Google Cloud environments |
+|  Service account key| App can self sign a google token with the service account key in order to accessing private data on behalf of a service account outside Google Cloud environments|
+
+
+
+## Useful links
+- [Feature doc](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+Authenticator#/)
+- [GCP authentication overview]( https://cloud.google.com/docs/authentication)
+- [Google metadata server](https://cloud.google.com/compute/docs/storing-retrieving-metadata#querying)
+
+## Issue description
+We want enable GCE hosts to authenticate Conjur with their google identity in order to retrieve secrets from Conjur.
+
+## Out of scope
+The only authentication strategy we are supporting is Environment-provided service account in google Compute engine.
+Other strategies are not supported: Service account key, OAuth 2.0 client, API key
+Other google components are not supported:  App Engine, GKE, Cloud Run, Cloud Functions
+
+## Solution
+We will add a new GCP authenticator to Conjur, which GCE hosts will can authenticate with by delivering their host Conjur identity and google identity as JWT token.
+
+As mentioned in the [feature doc](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+Authenticator#/), before any authentication 
+request is sent to Conjur, the admin will load the authenticator policy: 
+
+```yaml
+# policy id needs to match the convention `conjur/authn-azure/<service ID>`
+- !policy
+  id: conjur/authn-gcp/mygcp
+  body:
+  - !webservice
+      
+  - !group apps
+  
+  - !permit
+    role: !group apps
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+All the following configurations will have constants values and will not require any action from the customer:
+
+* **provider-uri** = `https://www.googleapis.com/oauth2/v1/certs` #represnts the address to fetch the public key in order to validate the JWT token
+* **mandatory-claims** = `iss, exp, aud, iat, nbf` #the list of the mandatory claims to validate in the JWT token
+* **iss** = `https://accounts.google.com` #the issuer claim we should validate in the JWT token
+* **aud** = `conjur` #the audience claim we should validate in the JWT token
+
+### GCP Resource Restrictions
+To authenticate with Conjur, GCP-specific fields will need to be provided in the host annotations of the Conjur host identity with the following options:
+
+* **project_id** - A customizable unique identifier for your project.
```

Comment:
is it string ? what are its limitations

Reply:
Projcet id is string, what do yo mean by limitations?

Reply:
I meant if the string should have any specific structure

Reply:
Project ID: A customizable unique identifier for your project.

The project ID is a unique, user-assigned ID that can be used by Google APIs. If you do not specify a project ID during project creation, a project ID will be generated automatically.

**The project ID must be a unique string of 6 to 30 lowercase letters, digits, or hyphens. It must start with a letter, and cannot have a trailing hyphen. You cannot change a project ID once it has been created.**

Some words are restricted from use in project IDs. If you use restricted words in the project name, such as google or ssl, the generated project ID will not include these words.
[Link](https://cloud.google.com/resource-manager/docs/creating-managing-projects)